### ISSUE TITLE
 TextBrewer/src/textbrewer/distiller_utils.py  get_outputs_from_batch fails tocheck dicts properly for maskedLM

### ISSUE THREAD

**Issue by AddedK** [airaria/TextBrewer#98]
* Models: base-multilingual-cased, 6 and 3 layers. MaskedLM models.
* Data: mBERT tokenized data for language modeling, using Swedish Gigacorpus or wikitext-2-raw-v1.
* Error: "str object has no attribute size" when the Transformer BERT code tries to get the shape of the input_ids.
I got the same bug whether or not I set `return_dict= False ` for both teacher and student models.
When I fork and modify the code, it seems that the batch has a type dict, even though there is a type check for it at
the beginning of the function.
The error happens for both the Swedish Gigacorpus and wikitext-2-raw-v1.

The following hack seems to work for when using the Swedish Gigacorpus:
```
def get_outputs_from_batch(batch, device, model_T, model_S, args, no_teacher_forward=False):
    if type(batch) is dict:
        .
        .
    else:
        .
        .
        else:
            with torch.no_grad():
                results_T = auto_forward(model_T,batch,args)
        print(type(batch)) # This prints <class 'dict'>
        # results_S = model_S(*batch, **args) # This original code does not work for me
        results_S = model_S(**batch, **args)  # This modification _does_ work
        teacher_batch = student_batch = batch
```
Here is the python code when trying to do task-agnostic distillation for the wikitext-2-raw-v1 dataset.
```
# -*- coding: utf-8 -*-
"""distillWikitext.ipynb

Automatically generated by Colaboratory.

"""

# !pip install transformers
# !pip install datasets
# !pip install textbrewer

import transformers

print(transformers.__version__)

import torch
SEED = 100 # Seeds to be used: 100, 101, 102
torch.manual_seed(SEED) 
device='cuda' if torch.cuda.is_available() else 'cpu'


block_size = 512


TRUNCATION = True # Affects how model is initialized
if TRUNCATION:
    SUFFIX = "truncated"
else:
    SUFFIX = "random"

DOING_VALIDATION = True
SMALL_MODE = True

if SMALL_MODE:
    NUM_LAYERS_TEACHER = 6
    NUM_LAYERS_STUDENT = 3
    SIZE_STRING = "SMALL"
else:
    NUM_LAYERS_TEACHER = 12
    NUM_LAYERS_STUDENT = 6
    SIZE_STRING = "FULL"

STUDENT_MODEL_NAME = "MBERT_6"  
TOKEN_MODEL = STUDENT_MODEL_NAME.split("_")[0]


OUTPUT_DIR = "distilled_GIGA_{0}_{1}_{2}_{3}".format(SIZE_STRING,SEED,STUDENT_MODEL_NAME,SUFFIX)
LOG_DIR = "distilled_GIGA_{0}_{1}_{2}_{3}_logs".format(SIZE_STRING,SEED,STUDENT_MODEL_NAME,SUFFIX)
MODEL_SAVE_FILE = "distilled_GIGA_{0}_{1}_{2}_{3}.pt".format(SIZE_STRING,SEED,STUDENT_MODEL_NAME,SUFFIX)


# Choose teacher

TEACHER_MODEL_NAME = "MBBERT_12" 
model_checkpoint = "bert-base-multilingual-cased"


NUM_EPOCHS = 3
WEIGHT_DECAY = 0.01

kbbert_6_train_args = {
    "lr": 1e-4, # candidates: 1e-4, 2e-4, 5e-5
    "batch_size": 2, # candidates: 2, 4
}
mbert_6_train_args = {
    "lr": 1e-4, # candidates: 1e-4, 2e-4, 5e-5
    "batch_size":  2, # candidates: 2, 4, 4 CUDA OUT OF MEMORY
}

if STUDENT_MODEL_NAME  == "KBBERT_6":
    train_args = kbbert_6_train_args
elif STUDENT_MODEL_NAME  == "MBERT_6":
    train_args = mbert_6_train_args
else:
    raise Exception("Undefined STUDENT_MODEL_NAME  {0}".format(STUDENT_MODEL_NAME ))


INTERMEDIATE_MATCHES = [    
        {'layer_T':NUM_LAYERS_TEACHER-1, 'layer_S':NUM_LAYERS_STUDENT-1, 'feature':'hidden', 'loss': 'hidden_mse','weight' : 1},
        {'layer_T':NUM_LAYERS_TEACHER-1, 'layer_S':NUM_LAYERS_STUDENT-1, 'feature':'attention', 'loss': 'attention_mse', 'weight' : 1}]

HARD_LABEL_WEIGHT = 0
KD_LOSS_WEIGHT = 0
TEMPERATURE = 4 # Doesn' matter, as we are ignoring logit loss anyway

from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
from transformers import DataCollatorForLanguageModeling
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm_probability=0.15)

from datasets import load_dataset
datasets = load_dataset('wikitext', 'wikitext-2-raw-v1')

def tokenize_function(examples):
    return tokenizer(examples["text"])

tokenized_datasets = datasets.map(tokenize_function, batched=True, num_proc=4, remove_columns=["text"])

def group_texts(examples):
    # Concatenate all texts.
    concatenated_examples = {k: sum(examples[k], []) for k in examples.keys()}
    total_length = len(concatenated_examples[list(examples.keys())[0]])
    # We drop the small remainder, we could add padding if the model supported it instead of this drop, you can
        # customize this part to your needs.
    total_length = (total_length // block_size) * block_size
    # Split by chunks of max_len.
    result = {
        k: [t[i : i + block_size] for i in range(0, total_length, block_size)]
        for k, t in concatenated_examples.items()
    }
    result["labels"] = result["input_ids"].copy()
    return result

lm_datasets = tokenized_datasets.map(
    group_texts,
    batched=True,
    batch_size=1000,
    num_proc=4,
)

tokenizer.decode(lm_datasets["train"][1]["input_ids"])

from transformers import BertConfig, AutoModelForMaskedLM, AutoConfig

# Download model from online, and save it's state dict on disk 
teacher_model = AutoModelForMaskedLM.from_pretrained(model_checkpoint)
teacher_config = AutoConfig.from_pretrained(model_checkpoint)

torch.save(teacher_model.state_dict(), 'mbert_teacher.pt')
teacher_config.to_json_file("mbert_teacher_config.json")

"""### Load teacher"""

teacher_config = BertConfig.from_json_file("mbert_teacher_config.json")
teacher_config.output_hidden_states = True
teacher_config.output_attentions = True

teacher_model = AutoModelForMaskedLM.from_config(teacher_config) 
teacher_model.load_state_dict(torch.load('mbert_teacher.pt'))
teacher_model.to(device=device)

"""### Prepare student"""

bert_config_T3 = BertConfig.from_json_file("mbert_teacher_config.json") # NOTE, using same config as teacher!
bert_config_T3.output_hidden_states = True
bert_config_T3.output_attentions = True
bert_config_T3.num_hidden_layers= 6 # Half as many layers

import torch.nn as nn
# Student model
TRUNCATION = True
if not TRUNCATION:
    # Since we are doing model constructor and not from_pretrained, student is randomly initialized
    student_model = AutoModelForMaskedLM.from_config(bert_config_T3) 
else:
    # Help from https://github.com/huggingface/transformers/issues/1206
    student_model = AutoModelForMaskedLM.from_pretrained(model_checkpoint,config=bert_config_T3)
    # It says that some models of distilbert were not used to initialized, layer 5,4,3, which are the top layers of the distilbert. Looks good!
student_model.to(device=device)

import textbrewer
from textbrewer import GeneralDistiller
from textbrewer import TrainingConfig, DistillationConfig
from transformers import BertConfig, AdamW,BertTokenizer, TrainingArguments
from transformers import get_linear_schedule_with_warmup

from torch.utils.data import DataLoader
train_data = lm_datasets["train"]
val_data = lm_datasets["validation"]

train_dataloader = DataLoader(train_data, shuffle=True, batch_size=4, collate_fn = data_collator) # Huge!
eval_dataloader = DataLoader(val_data, shuffle=True, batch_size=4, collate_fn = data_collator) # Huge!

BATCH_SIZE = 4
if DOING_VALIDATION:
    print("Fixing predict function for validation")
    def predict(model, val_data, device, step): # model=self.model_S, step = global_step mandatory keywords
        # https://huggingface.co/course/chapter7/2?fw=pt
        print("doing prediction")
        model.eval()

        args = TrainingArguments(
            OUTPUT_DIR,
            evaluation_strategy="no",
            save_strategy="no",
            learning_rate= train_args["lr"],
            num_train_epochs=3,
            weight_decay=0.01,
            do_train=False,
            do_eval=True,
            per_device_eval_batch_size = BATCH_SIZE
        )

        trainer = Trainer(
            model=model,
            args=args,
            data_collator=data_collator,
            eval_dataset = val_data,
            tokenizer=tokenizer,
        )

        eval_results = trainer.evaluate()
        print(eval_results['eval_loss'])
        del trainer
        return eval_results['eval_loss']
    from functools import partial
    callback_fun = partial(predict, model=student_model, eval_dataset = val_data, device=device) # fill other arguments
else:
    callback_fun = None

"""### Perform distillation"""

num_training_steps = len(train_dataloader) * NUM_EPOCHS 

optimizer = AdamW(student_model.parameters(), lr=train_args["lr"] )
scheduler_class = get_linear_schedule_with_warmup
scheduler_args = {'num_warmup_steps':int(0.1*num_training_steps), 'num_training_steps':num_training_steps}


def simple_adaptor(batch, model_outputs):
    return {'logits': model_outputs.logits, 'hidden': model_outputs.hidden_states, 'attention': model_outputs.attentions}
    

distill_config = DistillationConfig( # Mostly default config, except no logit
    temperature=TEMPERATURE,
    temperature_scheduler='none',
    hard_label_weight= HARD_LABEL_WEIGHT,  # NOTE!
    hard_label_weight_scheduler='none', 
    kd_loss_type='mse', 
    kd_loss_weight = KD_LOSS_WEIGHT,  # NOTE!
    kd_loss_weight_scheduler='none', 
    probability_shift=False,
    is_caching_logits=False,
    intermediate_matches=INTERMEDIATE_MATCHES)

train_config = TrainingConfig( # Default training configuration
    gradient_accumulation_steps=1,
    ckpt_frequency=1,
    ckpt_epoch_frequency=1, 
    ckpt_steps=None,
    log_dir=LOG_DIR,
    output_dir= OUTPUT_DIR,
    device=device,
    fp16=False,
    fp16_opt_level='O1',
    data_parallel=False,
    local_rank=-1)

distiller = GeneralDistiller(
    train_config=train_config, distill_config=distill_config,
    model_T=teacher_model, model_S=student_model, 
    adaptor_T=simple_adaptor, adaptor_S=simple_adaptor)

with distiller:
    distiller.train(optimizer, train_dataloader, NUM_EPOCHS ,
    scheduler_class=scheduler_class, scheduler_args = scheduler_args)
```

**Comment by airaria** [comment]
I found that the type of the batch is `transformers.tokenization_utils_base.BatchEncoding`.
So change the type checking `type(batch) is dict` to `isinstance(batch,abc.Mapping)`  in `distiller_utils.py` should fix this problem.

**Comment by airaria** [comment]
fixed in #100
